spec/models/course/registration_spec.rb

```diff
@@ -0,0 +1,108 @@
+require 'rails_helper'
+
+RSpec.describe Course::Registration, type: :model do
+  it { is_expected.to belong_to(:user) }
+  it { is_expected.to belong_to(:course) }
+  it { is_expected.to define_enum_for(:role) }
+
+  let!(:instance) { create(:instance) }
+  with_tenant(:instance) do
+    let!(:course) { create(:course) }
+    let!(:user) { create(:user) }
+    let!(:student) { create(:student_registration, course: course) }
+    let!(:teaching_assistant) { create(:teaching_assistant_registration, course: course) }
+    let!(:manager) { create(:manager_registration, course: course) }
+    let!(:owner) { create(:owner_registration, course: course) }
+
+    context 'registering the same user to the same course twice' do
+      subject { create(:course_registration, course: course, user: user, role: :student) }
+      before do
+        create(:course_registration, course: course, user: user, role: :student)
+      end
+      it 'fails' do
+        expect { subject }.to change(Course::Registration, :count).by(0).and raise_error
+      end
+    end
+
+    describe '.student' do
+      it 'returns student registrations' do
+        expect(course.registrations.student).to contain_exactly(student)
+      end
+    end
+
+    describe '.staff' do
+      it 'returns teaching assistant, manager and owner registrations' do
+        expect(course.registrations.staff).to contain_exactly(teaching_assistant, manager, owner)
+      end
+    end
+
+    describe '.process' do
+      before do
+        @registration_ids = [student.id, teaching_assistant.id, manager.id]
+        @approve = true
+        @process_all = false
+        @role = nil
+      end
+
+      subject { Course::Registration.process!(course, @registration_ids, approve: @approve,
```

Comment:
Avoid using {...} for multi-line blocks.<br>Block body expression is on the same line as the block start.